glommio/src/sys/uring.rs

```diff
@@ -731,29 +732,56 @@ trait UringCommon {
         completed
     }
 
-    /// It is important to process cancellations as soon as we see them,
-    /// which is why they go into a separate queue. The reason is that
-    /// cancellations can be racy if they are left to their own devices.
+    /// Used when we absolutely need to flush at least min_requests requests
+    /// from queue. This function blocks if it has to.
     ///
-    /// Imagine that you have a write request to fd 3 and wants to cancel it.
-    /// But before the cancellation is run fd 3 gets closed and another file
-    /// is opened with the same fd.
-    fn flush_cancellations(&mut self, woke: &mut usize) {
-        let mut cnt = 0;
+    /// min_requests does not account for SQEs waiting to be sent to the kernel.
+    /// If we have 2 items in the queue and 8 SQEs pending submission then
+    /// min_requests = 1 will submit all 8 SQEs + the first item in the queue.
+    ///
+    /// The function returns the number of elements submitted. This can be lower
+    /// than min_requests if there were fewer requests in the queue.
+    fn force_flush_ring(
+        &mut self,
+        queue: &mut VecDeque<UringDescriptor>,
+        mut min_requests: usize,
+        woke: &mut usize,
+    ) -> usize {
+        // we can't submit more request than what's there to submit + pending SQEs
+        min_requests = min_requests.min(queue.len()) + self.waiting_kernel_submission();
+
+        // if we fail to submit we need to make sure we collect CQEs before sleeping
+        let mut consumed = false;
+        let mut submitted = 0;
         loop {
-            if self.consume_cancellation_queue().is_ok() {
-                break;
+            match self.consume_sqe_queue(queue, true) {
+                Err(err) => {
+                    if !consumed {
+                        consumed = true;
+                    } else {
+                        warn!(
+                            "failed to flush the required minimum of events ({}/{}): {}; waiting \
+                             for {} CQEs",
+                            submitted,
+                            min_requests,
+                            err,
+                            min_requests - submitted
+                        );
+                        self.wait_for_events((min_requests - submitted) as u32)
+                            .expect("failed to wait for CQE. Game over");
+                        consumed = false;
+                    }
+                    self.consume_completion_queue(woke);
+                }
+                Ok(x) => submitted += x,
             }
-            self.consume_completion_queue(woke);
```

Comment:
I feel that this still deserves at least a `warn!()`. If you have executed this loop 1MM times... at the very least you have a giant performance problem on your hands and I bet you'd love to know where it came from

Reply:
See the warning here https://github.com/DataDog/glommio/pull/512/files#diff-d8d2d815ed457d48361ae92c5976d152059d3f04d59dea4ad148d1ea8a75f21aR761-R768